roles/lib_openshift/src/class/oc_adm_csr.py

```diff
@@ -0,0 +1,194 @@
+
+            for ocnode in results:
+                if node in ocnode['metadata']['name']:
+                    nodes_list[-1]['accepted'] = True
+
+        return nodes_list
+
+    def get(self):
+        '''get the current certificate signing requests'''
+        return self.csrs
+
+    @staticmethod
+    def action_needed(csr, action):
+        '''check to see if csr is in desired state'''
+        if csr['status'] == {}:
+            return True
+
+        state = csr['status']['conditions'][0]['type']
+
+        if action == 'approve' and state != 'Approved':
+            return True
+
+        elif action == 'deny' and state != 'Denied':
+            return True
+
+        return False
+
+    def match_node(self, csr):
+        '''match an inc csr to a node in self.nodes'''
+        for node in self.nodes:
+            # we have a match
+            if node['name'] in csr['metadata']['name']:
+                node['csrs'][csr['metadata']['name']] = csr
+
+                if (node['name'] in csr['spec']['username'] and
+                        csr['status'] and
+                        csr['status']['conditions'][0]['type'] == 'Approved'):
+                    node['accepted'] = True
+
+                return node
+
+        return None
+
+    def finished(self):
+        '''determine if there are more csrs to sign'''
+        # if nodes is set and we have nodes then return if all nodes are 'accepted'
+        if self.nodes is not None and len(self.nodes) > 0:
+            return all([node['accepted'] for node in self.nodes])
+
+        # we are approving everything or we still have nodes outstanding
+        return False
+
+    def manage(self, action):
+        '''run openshift oc adm ca create-server-cert cmd and store results into self.nodes
+
+           we attempt to verify if the node is one that was given to us to accept.
+
+           action - (allow | deny)
+        '''
+
+        results = []
+        # There are 2 types of requests:
+        # - node-bootstrapper-client-ip-172-31-51-246-ec2-internal
+        #   The client request allows the client to talk to the api/controller
+        # - node-bootstrapper-server-ip-172-31-51-246-ec2-internal
+        #   The server request allows the server to join the cluster
+        # Here we need to determine how to approve/deny
+        # we should query the csrs and verify they are from the nodes we thought
+        for csr in self.csrs:
+            node = self.match_node(csr)
+            # oc adm certificate <approve|deny> csr
+            # there are 3 known states: Denied, Aprroved, {}
+            # verify something is needed by OCcsr.action_needed
+            # if approve_all, then do it
+            # if you passed in nodes, you must have a node that matches
+            if self.approve_all or (node and OCcsr.action_needed(csr, action)):
+                result = self.openshift_cmd(['certificate', action, csr['metadata']['name']], oadm=True)
```

Comment:
@ashcrow, I'll try and explain.  I'm not in love with this method of approving but I'm not sure how else to accomplish this task.

The node first comes up with a kubeconfig placed on the host through cloud user_data.  The host uses the kubeconfig (token) to talk to the master API/controllers.  
1. This first node request is from the service account to talk to the API from the node.  Inside of the csr (oc get csr) on the master it returns a field named "username".  This "username" field has the service account in it.  The admin calls `oc adm certificate approve/deny <csr>` and the certificate is approved or denied.
2. Once the service account is approved to talk to the API/controllers, a joining request is issued to the node.  The second csr that comes into the master is for the node itself to join the cluster.  The "username" field has the name of the node in it.  Thus, line 128 must match the username in order for us to consider the node accepted only after both 1 and 2 have been approved.

Its kind of confusing that there are 2 csr's per node.  Only after the 2nd one is approved do we consider the node `node['accepted']=True`.

Hopefully this sheds some light on it.  If not, we can walk through it tomorrow and I can demonstrate it better.

If we deny the 1st request then it will never issue the 2nd.  I can't think of a scenario where you would allow the server API access but then deny the node into the cluster.  It would probably be a security issue if this were to happen.  Once denied I was unable to then re-approve the node.  It never exited the denied state.